### DISCUSSION TITLE
To have a curated and garbage-free history

### DISCUSSION THREAD

**Discussion by sassanh** [fish-shell/fish-shell#10785] [Ideas]
Hi guys,
I really appreciate your work on this wonderful open source product!

I would like to talk about an issue here, I don't have a solution for it myself, but the problem is real.

I, like many others work on different projects with different stacks, each including different tech, potentially with commands and each command exposing their own cli.

Some of these commands follow unix best practices for cli, while some of them are very cryptic with hard to remember options and arguments. They mostly make sense when you are in the context, but as you switch context, they lose their meaning and it becomes hard to remember.

Now for each of these commands, at some point I may spend some time to learn them deeply and find the options and arguments in their docs and possibly with some trial and error, at this point everything makes sense and I know exactly how it behaves, but when I return to it after say 2 weeks, it becomes all alien language again. Now one very helpful thing to get me back in the context, is my fish history, I write some keywords and hit "k" (I'm in vim mode) and see previous commands and it really helps.

The thing is during the trial and error phase, I probably executed many wrong commands and they are all also in my history. If it makes sense for a particular command to run with `--remote` and after half an hour of trial and error I understand `--remote` is not relevant and shouldn't be used in my case, after a month, odds are high that I forget about it and start searching my history with `remote` and get all the wrong commands again and try them again until I find out I should search some other keyword.

At this point, I hope I was able to communicate the issue. I think terminal shells can probably provide tools to help with this issue.

One solution with current tooling is to run all the commands in trial and error phase in an incognito shell and then come back to a normal shell and run the best version of the command to record it in the history. But it's not easy as you need to be conscious in which terminal you are trying and make sure you write the good command at least once in a normal shell, and then you may find out what you thought was right a minute ago and recorded it, is actually wrong. (using clear-session is an alternative of incognito terminal, with similar challenges.)

Another solution would be an interactive history (like browser history) in which you can easily select bunch of history items and remove them. Editing fish_history in an editor can be a similar experience. This approach is very manual and very precise. You have total control.

Another solution would be to set the terminal in a "trial-and-error" mode, in which it records no commands unless you hit a specific key combination after it's been launched and it renders an indicator for recorded commands and a different indicator for commands that haven't been recorded yet.

It would be nice if one could also record commands before the last one, one solution for this would be to have the indicator that renders next to unsaved commands, show an index too, so for example the 15th previous unsaved command, shows index 5 and to save it, I need to hit the key combination with the number 15 somewhere in it (like vim's <count> in its mappings). This is a bit like iPython's `In[15]`/`Out[15]` user experience.

**Comment by krobelus** [comment]
There are several ways to delete items from history:
- the upcoming release has this (just like your browser)
> - During up-arrow history search, :kbd:`shift-delete` will delete the current search item and move to the next older item. Previously this was only supported in the history pager.
>   Same for autosuggestions.
- `history delete`
- `fish_config` (the web interface and press X next to a command)

and we also have some ways of preventing commands from being added to the (default) history
- prefix the command with a space. Then it will only be saved in memory, great for experimentation.
- the upcoming release has
> - A new function ``fish_should_add_to_history`` can be overridden to decide whether a command should be added to the history (:issue:`10302`).
- use `set fish_history tmp_history` to switch to a different history file and `set -e fish_history` to switch back
- use `fish_private_mode`

**Reply by sassanh** [reply]
Thanks for listing all this options!
I wasn't aware of the web interface having a feature to delete history items, I also wasn't aware of the new features in the upcoming changes, I think the combination of these features are more than enough to curate history neat and clean. I already love this `shift-delete` key when I read about it.
